File: python/7.StacksAndQueues/StoneWall.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")
from collections import deque

def solution(H):
    stack = deque()
    count = 1
    stack.append(H[0])
    for i in range (1, len(H)):
        while len(stack) > 0 and stack[-1] > H[i]:
            stack.pop()
        if len(stack) == 0 or H[i] > stack[-1]:
            stack.append(H[i])
            count += 1
    return count

File: csharp/7.StacksAndQueues/StoneWall.cs
using System;
// you can also use other imports, for example:
// using System.Collections.Generic;
using System.Collections.Generic;
// you can write to stdout for debugging purposes, e.g.
// Console.WriteLine("this is a debug message");

class Solution {
    public int solution(int[] H) {
        Stack<int> stack = new Stack<int>();
        int count = 1;
        stack.Push(H[0]);
        for (int i = 1; i < H.Length; i++)
        {
            int block = H[i];

            while (stack.Count > 0 && stack.Peek() > block)
            {
                stack.Pop();
            }

            if (stack.Count == 0 || block > stack.Peek())
            {
                stack.Push(block);
                count++;
            }

        }
        return count;
    }
}

